File: leetcode/#190.py
# question --> https://leetcode.com/problems/reverse-bits/

class Solution:
    def reverseBits(self, n: int) -> int:
        q, r = n, 0
        binA = ''
        
        while q >= 1: 
            r = q % 2
            q //= 2
            binA += str(r)
            
        zeroes = 32 - len(binA)
        for _ in range(zeroes):
            binA += '0'
            
        i, j, ans = 31, 0, 0
        while j < 32: 
            if binA[j] == '1':
                ans += pow(2, i)
                
            i -= 1
            j += 1
            
        return ans
        
        # TC: O(logN); SC: O(logN)
        
# end 

File: leetcode/#190.js
// question --> https://leetcode.com/problems/reverse-bits/

/**
 * @param {number} n - a positive integer
 * @return {number} - a positive integer
 */
var reverseBits = function(n) {
    let q = n; 
    let r = 0;
    let binA = '';
    
    while (q >= 1) {
        r = q % 2;
        q = Math.floor(q / 2);
        binA += r.toString(); 
    }
    
    let zeros = 32 - binA.length;
    for (let i = 0; i < zeros; i++) {
        binA += '0';
    }
    
    let x = 31; 
    let j = 0; 
    let ans = 0; 
    while (j < 32) {
        if (binA[j] == '1') {
            ans += Math.pow(2, x);
        }
        x -= 1;
        j += 1;
    }
    
    return ans;
};

// TC: O(logN); SC: O(logN)

// end